tests/types/test_numpy.py

```diff
@@ -0,0 +1,379 @@
+
+    cur = numpy_conn.cursor()
+
+    cur.execute(f"select pg_typeof({expr}) = pg_typeof(%{fmt_in.value})", (val,))
+    cur.fetchone()[0] is True
+
+    cur.execute(f"select {expr} = %{fmt_in.value}", (val,))
+    assert cur.fetchone()[0] is True
+
+
+@pytest.mark.parametrize(
+    "val, expr",
+    [
+        (0, "'0'::numeric"),
+        (18_446_744_073_709_551_615, "'18446744073709551615'::numeric"),
+    ],
+)
+@pytest.mark.parametrize("fmt_in", PyFormat)
+def test_dump_numpy_ulonglong(numpy_conn, val, expr, fmt_in):
+
+    val = np.ulonglong(val)
+
+    assert isinstance(val, np.ulonglong)
+
+    cur = numpy_conn.cursor()
+
+    cur.execute(f"select pg_typeof({expr}) = pg_typeof(%{fmt_in.value})", (val,))
+    cur.fetchone()[0] is True
+
+    cur.execute(f"select {expr} = %{fmt_in.value}", (val,))
+    assert cur.fetchone()[0] is True
+
+
+# Test float special values
+@pytest.mark.parametrize(
+    "val, expr",
+    [
+        (np.PZERO, "'0.0'::float8"),
+        (np.NZERO, "'-0.0'::float8"),
+        (np.nan, "'NaN'::float8"),
+        (np.inf, "'Infinity'::float8"),
+        (np.NINF, "'-Infinity'::float8"),
+    ],
+)
+@pytest.mark.parametrize("fmt_in", PyFormat)
+def test_dump_special_values(numpy_conn, val, expr, fmt_in):
+    # All numpy special values are python float types,
+    # So the Dumper used should be the normal float dumper
+
+    if val == np.nan:
+        assert np.nan == np.NAN == np.NaN
+
+    if val == np.inf:
+        assert np.inf == np.Inf == np.PINF == np.infty
+
+    assert isinstance(val, float)
+
+    cur = numpy_conn.cursor()
+    cur.execute(f"select pg_typeof({expr}) = pg_typeof(%{fmt_in.value})", (val,))
+
+    assert cur.fetchone()[0] is True
+
+    cur.execute(f"select {expr} = %{fmt_in.value}", (val,))
+    assert cur.fetchone()[0] is True
+
+
+@pytest.mark.parametrize(
+    "val, expr",
+    [
+        (31e3, "31e3::float4"),
+        (31e-3, "31e-3::float4"),
+    ],
+)
+@pytest.mark.parametrize("fmt_in", PyFormat)
+def test_dump_numpy_float16(numpy_conn, val, expr, fmt_in):
+
+    val = np.float16(val)
+
+    cur = numpy_conn.cursor()
+
+    cur.execute(f"select pg_typeof({expr}) = pg_typeof(%{fmt_in.value})", (val,))
+    cur.fetchone()[0] is True
+
+    cur.execute(
+        f"select abs(({expr}::float4 - %{fmt_in.value}) / {expr}::float4)", (val,)
+    )
+    res = cur.fetchone()[0]
+
+    assert res <= 1e-3
+
+
+@pytest.mark.parametrize(
+    "val, expr",
+    [
+        (256e6, "256e6::float4"),
+        (256e-6, "256e-6::float4"),
+    ],
+)
+@pytest.mark.parametrize("fmt_in", PyFormat)
+def test_dump_numpy_float32(numpy_conn, val, expr, fmt_in):
+
+    val = np.float32(val)
+
+    cur = numpy_conn.cursor()
+
+    cur.execute(f"select pg_typeof({expr}) = pg_typeof(%{fmt_in.value})", (val,))
+    cur.fetchone()[0] is True
+
+    cur.execute(
+        f"select abs(({expr}::float4 - %{fmt_in.value}) / {expr}::float4)", (val,)
+    )
+    res = cur.fetchone()[0]
+
+    assert res <= 1e-6
```

Comment:
This test doesn't convince me.

We are testing the dumper, so what we want to be sure is that in the leg from the numpy.XXX type to postgres we don't change the data we are dumping. As such I don't think it's a good idea to do arithmetic, and it wouldn't even be correct to select the value in Python and test the result, because we would test the loading part too. So, the test should be in postgres itself.

I expect us to be able to write test that verify that we don't add approximation to floating point values, so we should check that `a = b` in postgres, which is not what you usually do with floats (you test that |a - b| < Ɛ), but here we really want to be bit-precise, if we can.

For the same reason, I'd express the value in the fixtures as string instead of numbers, so we never pass through the Python float type and its approximations.

How about the following? I also added a few other values in the samples, which are obtained by printing the repr of the numpy values (`repr(numpy.float32(numpy.e))`)


```suggestion
@pytest.mark.parametrize(
    "val",
    [
        "256e6",  # a large value
        "256e-6",  # a small value
        "2.7182817",  # e
        "3.1415927",  # pi
    ],
)
@pytest.mark.parametrize("fmt_in", PyFormat)
def test_dump_numpy_float32(numpy_conn, val, fmt_in):

    val = np.float32(val)

    cur = numpy_conn.cursor()

    cur.execute(f"select pg_typeof({val}::float4) = pg_typeof(%{fmt_in.value})", (val,))
    assert cur.fetchone()[0] is True

    cur.execute(f"select {val}::float4 = %{fmt_in.value}", (val,))
    assert cur.fetchone()[0] is True
```